File: scripts/icons/src/manifest.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct Manifest {
    pub author: String,
    pub manifest_version: i64,
    pub browser_action: BrowserAction,
    pub background: Background,
    pub icons: HashMap<String, String>,
    pub permissions: Vec<String>,
    pub content_scripts: Vec<ContentScript>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Background {
    pub scripts: Vec<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BrowserAction {
    pub default_popup: String,
    pub default_title: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ContentScript {
    pub matches: Vec<String>,
    pub js: Vec<String>,
}

File: src/manifest.ts
/* eslint-disable import/no-extraneous-dependencies */
import fs from 'fs-extra';
import type { Manifest } from 'webextension-polyfill';
import type PkgType from '../package.json';
import { isDev, r } from '../scripts/utils';

// eslint-disable-next-line import/prefer-default-export
export async function getManifest() {
  const pkg = (await fs.readJSON(r('package.json'))) as typeof PkgType;

  // update this file to update this manifest.json
  // can also be conditional based on your need
  const manifest: Manifest.WebExtensionManifest = {
    manifest_version: 2,
    version: '1.2.1',
    description:
      'Instantly check which of your favorite Twitch streamers are live right from your favorite browser',
    name: 'Now Live',
    browser_action: {
      default_popup: 'dist/popup/index.html',
      default_title: 'See who is live',
    },
    background: {
      scripts: ['dist/background/index.mjs'],
    },
    icons: {
      16: 'assets/icons/16.png',
      32: 'assets/icons/32.png',
      48: 'assets/icons/48.png',
      64: 'assets/icons/64.png',
      96: 'assets/icons/96.png',
      128: 'assets/icons/128.png',
      256: 'assets/icons/256.png',
    },
    permissions: ['storage', 'alarms', '*://*.twitch.tv/*'],
    content_scripts: [
      {
        matches: ['*://nowlive.jewelexx.com/auth/callback'],
        js: ['dist/contentScripts/index.global.js'],
      },
    ],
  };

  // FIXME: not work in MV3
  if (isDev && false) {
    // for content script, as browsers will cache them for each reload,
    // we use a background script to always inject the latest version
    // see src/background/contentScriptHMR.ts
    delete manifest.content_scripts;
    manifest.permissions?.push('webNavigation');
  }

  return manifest;
}
